homie/src/lib.rs

```diff
@@ -594,6 +605,31 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn disconnect_succeeds_before_ready() -> Result<(), SendError<Request>> {
+        let (mut device, rx) = make_test_device();
+
+        device.start().await?;
+        device.disconnect().await?;
+
+        // Need to keep rx alive until here so that the channel isn't closed.
+        drop(rx);
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn disconnect_succeeds_after_ready() -> Result<(), SendError<Request>> {
+        let (mut device, rx) = make_test_device();
+
+        device.start().await?;
+        device.ready().await?;
+        device.disconnect().await?;
+
+        // Need to keep rx alive until here so that the channel isn't closed.
+        drop(rx);
```

Comment:
It's just occurred to me that we could read from rx and assert the events match some snapshot.

Reply:
We could do, but it would end up being a lot of events, and would make the tests quite brittle as it would verify the order of events rather than the final state. I think what I'd rather do is have some kind of integration test where we spin up a local MQTT server and (once I've implemented it) the Homie controller library, and make sure that it sees the expected state.